Scroll-driven animations: animation-range and <timeline-range-name> data type (#42020)
mdn/content: files/en-us/web/css/reference/properties/animation-range/index.md

## Syntax

```css
/* Single value syntax */
animation-range: normal;
animation-range: 20%;
animation-range: 100px;
animation-range: cover;
animation-range: contain;

/* Two value syntax */
/* Named timeline range value */
animation-range: cover 20%;
animation-range: contain 100px;
/* Range start and end */
animation-range: normal 25%;
animation-range: 25% normal;
animation-range: 25% 50%;
animation-range: entry exit;

/* Three value syntax */
animation-range: cover cover 200px;
animation-range: 10% exit 90%;
animation-range: entry 10% 90%;

/* Four value syntax */
animation-range: cover 0% cover 200px;
animation-range: entry 10% exit 100%;
```

### Values

> - `<animation-range-start>`

Author: added under syntax

  - : The keyword `normal`, a `<length-percentage>`, a [`<timeline-range-name>`](/en-US/docs/Web/CSS/Reference/Values/timeline-range-name), or a `<timeline-range-name> <length-percentage>` pair, representing the [`animation-range-start`](/en-US/docs/Web/CSS/animation-range-start). If a `<timeline-range-name>` is set without a `<length-percentage>`, the `<length-percentage>` defaults to `0%`.
- `<animation-range-end>`
  - : The keyword `normal`, a `<length-percentage>`, a `<timeline-range-name>`, or a `<timeline-range-name> <length-percentage>` pair, representing the [`animation-range-end`](/en-US/docs/Web/CSS/animation-range-end). If a `<timeline-range-name>` is set without a `<length-percentage>`, the `<length-percentage>` defaults to `100%`.

## Description

> The `animation-range` shorthand property sets the `animation-range-start` and `animation-range-end` values, defining where along the animation timeline the animation will start and end. The `0%` keyframe effect will not be applied to the element until the range start is reached if the {{cssxref("animation-fill-mode")}} is not set to `backwards` or `both`.

Reviewer: ```suggestion
The `animation-range` shorthand property sets the `animation-range-start` and `animation-range-end` values, defining where along the animation timeline the animation will start and end. The `0%` keyframe effect will not be applied to the element until the range start is reached unless the {{cssxref("animation-fill-mode")}} is set to `backwards` or `both`.
```

Does that work? The double negative was confusing.

> The property can be applied to a container element as a combination of the `<animation-range-start>` and `<animation-range-end>` values. If both the values are specified, they will be interpreted in the order `<animation-range-start>` then `<animation-range-end>`.

Reviewer: That first sentence was confusing, I thought.

The value of each component is either the keyword `normal`, a {{cssxref("length-percentage")}}, or a [`<timeline-range-name>`](/en-US/docs/Web/CSS/Reference/Values/timeline-range-name), optionally defined with a `<length-percentage>`

> If the value is a {{cssxref("length-percentage")}} or `normal`, this is the value for the `<animation-range-start>` and `<animation-range-end>` will equal `normal`. For example, `animation-range: 20%;` is the equivalent to `animation-range: 20% normal;` and `animation-range: normal;` is the equivalent to `animation-range: normal normal;`.

Reviewer: ```suggestion
If the value is a {{cssxref("length-percentage")}} or `normal`, this is interpreted as the value for `<animation-range-start>`. The `<animation-range-end>` value will be implicitly set to `normal`. For example, `animation-range: 20%;` is equivalent to `animation-range: 20% normal;` and `animation-range: normal;` is equivalent to `animation-range: normal normal;`.
```

If two values are specified as components of the `<animation-range>` property, they will be interpreted in the order `<animation-range-start>` then `<animation-range-end>`. The value of each component is either the keyword `normal` or a {{cssxref("length-percentage")}} or a [`<timeline-range-name>`](/en-US/docs/Web/CSS/Reference/Values/timeline-range-name), optionally defined with a `<length-percentage>`. These values are space-separated. Normal is equal to `0%` for start and `100%` for end. Setting `normal` with a `<length-percentage>` for either the start or end range is invalid.

### Defining range start and defaulting range end

If the value is a {{cssxref("length-percentage")}} or `normal`, this is interpreted as the value for `<animation-range-start>`. The `<animation-range-end>` value will be implicitly set to `normal`. For example:

- `animation-range: 20%;` is the equivalent to `animation-range-start: 20%; animation-range-end: normal;`
- `animation-range: normal;` is the equivalent to `animation-range-start: 0%; animation-range-end: 100%;`

> If the value is a [`<timeline-range-name>`](/en-US/docs/Web/CSS/Reference/Values/timeline-range-name) without a `<length-percentage>` following it, that timeline range name is applied to the `<animation-range-start>` and `<animation-range-end>` components at `0%` and `100%`, respectively. For example:

Reviewer: Same for the next one? Maybe put those under a different heading?

## Examples

### Basic usage of the `animation-range` property

In this example, we reduce the duration of the view progress scroll animation by using the `animation-range` property to offset the start and end of the animation, and demonstrate the effect of the {{cssxref("animation-fill-mode")}} property on shortened animation timelines.

#### HTML

> In the middle of a wall of text, we include an element that we will animate. We added a lot of text in to ensure that our content overflows its container, but this was hidden for the sake of brevity.

Reviewer: ```suggestion
In the middle of a wall of text, we include an element that we will animate. We include a lot of text to ensure that our content overflows its container, but this is hidden for the sake of brevity.
```

    imperdiet sed euismod nisi. Eget egestas purus viverra accumsan in nisl nisi
    scelerisque. Netus et malesuada fames ac.
  </p>
  <p>
    <label>
      <input type="checkbox"> Add <code>animation-fill-mode: both;</code>
    </label>
  </p>
</div>
```

#### CSS

We define an animation that animates an element's opacity, scale, and color, causing it to fade in, size up, and change colors as the animation progresses. We apply this animation to the `animatedElement` with the {{cssxref("animation")}} shorthand.

> A view progress timeline is created by setting the [`view()`]() function as the value of the {{cssxref("animation-timeline")}} property on our `animatedElement`. The result is that the element animates as it moves upwards through the document as it is scrolled. We declared the shorthand before the `animation-timeline`, as the shorthand resets this property.

Reviewer: ```suggestion
A view progress timeline is created by setting the [`view()`]() function as the value of the {{cssxref("animation-timeline")}} property on our `animatedElement`. This results in the element animating as it moves upwards through the document when scrolled. We declared the shorthand before the `animation-timeline`, as the shorthand resets this property.
```

Lastly, an `animation-range` declaration is also set to make the animation begin later than expected, and finish earlier.

it moves up the scroller.

```css
.animatedElement {
  background-color: deeppink;
  animation: appear 1ms linear;
  animation-timeline: view();
  animation-range: entry 10% exit -25%;
}

@keyframes appear {
  from {
    background-color: rebeccapurple;
    opacity: 0;
    transform: scaleX(0);
  }

  to {
    background-color: darkturquoise;
    opacity: 0.75;
    transform: scaleX(0.75);
  }
}
```

> The `animatedElement` and its container's other styles were hidden for the sake of brevity.

Reviewer: ```suggestion
The other styles are hidden for the sake of brevity.
```
